Epic 9946: Ability to "hide" certain peptides/small molecules (AutoQC)

labkey-bpatel opened the pull request:
#### Rationale
Ability to hide certain peptide/small molecules in AutoQC plot page. This feature will help the scientist to filter the “good” peptides/molecule list, but still allow them to view the rest. 

#### Related Pull Requests
* https://github.com/LabKey/PanoramaPremiumModules/pull/77

#### Changes
* Custom queries and metadata for Peptide/Molecule precursors that groups precursors from different skyline documents
* Server side updates to only show Included precursors. 
* Add 'Show Excluded Precursors' checkbox to the plot panel.
* Schema update to add 'ExcludedPrecursors' table.

labkey-bpatel committed:
Create new lookup table for QC Group & populate with 'Included' & 'Excluded' values.
Update Molecule & Peptide schema to include QCGroupId column with default values from QC Group table.

labkey-bpatel committed:
Reformat code

labkey-bpatel committed:
- Custom queries for Peptide/Molecule precursors that groups precursors from different skyline documents.
- Action to mark Peptide/Molecule precursors as Excluded or Included.
- Targetedms schema update for holding excluded peptides/molecules.
- Updates to get the selected rows.

labkey-bpatel committed:
Updates for precursor identifier. Add molecule precursor metadata.

labkey-bpatel committed:
Server side updates to only show Included precursors. Add checkbox to the plot panel.

labkey-bpatel committed:
Add 'showExcludedPrecursors' to config

labkey-bpatel committed:
B
B
B
Merge remote-tracking branch 'origin/develop' into 22.3_fb_ability_to_hide_peptides

labkey-bpatel committed:
Schema update

labkey-bpatel committed:
Update precursor queries to round double values. Handle blank/bracketed ionFormula.

labkey-jeckels commented on resources/queries/targetedms/QCGroupingPrecursorMolecules.sql:
```diff
@@ -0,0 +1,40 @@
+        END) AS markedAs,
+FROM
+     (SELECT
+        ((CASE WHEN molPrec.MoleculeId.customIonName IS NOT NULL THEN molPrec.MoleculeId.customIonName
```
Unfortunately, you can have any combination of null and non-null values for these two columns. We should include both if they're non-null

labkey-bpatel replied:
done

labkey-jeckels commented on resources/queries/targetedms/QCGroupingPrecursorMolecules.sql:
```diff
@@ -0,0 +1,40 @@
+        molPrec.massMonoisotopic,
+        molPrec.massAverage
+        FROM
+            targetedms.Molecule mol
```
This query will be more efficient if you just use the MoleculePrecursor table, and use the lookup syntax to get to its MoleculeId instead of doing an explicit join here. That way we won't have to separately filter both tables on Container

labkey-bpatel replied:
done

labkey-jeckels commented on resources/queries/targetedms/QCGroupingPrecursorPeptides.sql:
```diff
@@ -0,0 +1,34 @@
+        prec.neutralMass,
+        prec.IsotopeLabelId.Name
+        FROM
+        targetedms.Peptide pep
```
Same thing on this join - if you just use Precursor and get to the peptide columns via the lookup syntax it'll be faster.

labkey-jeckels commented on resources/queries/targetedms/QCGroupingPrecursorPeptides.sql:
```diff
@@ -0,0 +1,34 @@
+        LEFT JOIN
+        targetedms.Precursor prec
+        ON
+        pep.PeptideModifiedSequence = prec.PeptideId.modifiedPeptideDisplayColumn
```
While this ON clause will go away based on my suggestion above, we should do joins on the Id columns and not the peptide sequence columns.

labkey-bpatel replied:
That makes sense

labkey-jeckels commented on src/org/labkey/targetedms/outliers/OutlierGenerator.java:
```diff
@@ -312,18 +318,28 @@ private void retainIfPreferred(Map<String, QCMetricConfiguration> preferredConfi
         DecimalFormat format = new DecimalFormat();
         format.setMinimumFractionDigits(4);
 
+        Collection<Map<String, Object>> excludedPrecursors = new TableSelector(schema.getTable("ExcludedPrecursors")).getMapCollection();
```
Please create a simple bean class instead of using a map.

labkey-bpatel replied:
done

labkey-jeckels commented on src/org/labkey/targetedms/outliers/OutlierGenerator.java:
```diff
@@ -312,18 +318,28 @@ private void retainIfPreferred(Map<String, QCMetricConfiguration> preferredConfi
+                        rs.getInt("Charge"),
+                        rs.getDouble("MZ")))
+                {
+                    excludedPrecursorsIds.put(rs.getLong("Id"), null);
```
Why is excludedPrecursorIds a map? If the value is always null, please use a Set.

labkey-bpatel replied:
Using a bean class, so no longer using this approach

labkey-jeckels commented on src/org/labkey/targetedms/outliers/OutlierGenerator.java:
```diff
@@ -344,6 +371,24 @@ private void retainIfPreferred(Map<String, QCMetricConfiguration> preferredConfi
 
+    private boolean isExcludedPrecursorPeptide(Collection<Map<String, Object>> excludedPrecursors, String modifiedSeq, int charge, double mz)
+    {
+        String peptidePrecursorIdentifier = modifiedSeq + "," + charge + "," + DECIMAL_FORMAT.format(mz);
```
This seems fragile since it needs to match up with how the SQL query is building the string. Can we find a way to consolidate implementations and/or add comments saying that they have to be kept in sync.

labkey-bpatel replied:
Now re-implemented.

labkey-jeckels commented on resources/schemas/dbscripts/postgresql/targetedms-22.001-22.002.sql:
```diff
@@ -0,0 +1,13 @@
+
+    CONSTRAINT PK_EXCLUDEDPRECURSORS_ROWID PRIMARY KEY (rowId),
+    CONSTRAINT FK_EXCLUDEDPRECURSORS_CONTAINER FOREIGN KEY (Container) REFERENCES core.Containers(EntityId),
+    CONSTRAINT UQ_EXCLUDEDPRECURSORS_PRECURSOR_IDENTIFIER UNIQUE (precursorIdentifier)
```
This should include Container, as you might exclude the same precursor in two different folders.

labkey-bpatel committed:
Code review changes:
- Use Precusor and MoleculePrecursor table instead of explicitly joining with Peptide and Molecule tables respectively.
- Include both customIonName and ionFormula if both are non-null.
- Create a simple bean class for ExcludedPrecusor instead of using a Map.
- Add/update comments.

labkey-bpatel committed:
Update targetedms schema to add precursor columns instead of using concatenated string of precursor info as a key.
Update custom queries for peptide and molecule precursors.
Update peptide and molecule precursor comparision that handles the exclusion in QC Plot.

labkey-bpatel committed:
Merge remote-tracking branch 'origin/release22.3-SNAPSHOT' into 22.3_fb_ability_to_hide_peptides

labkey-bpatel committed:
Update peptide query to use modifiedSequence.

labkey-jeckels commented on resources/queries/targetedms/QCGroupingPrecursorMolecules.sql:
```diff
@@ -0,0 +1,42 @@
+            molPrec.massAverage,
+            molPrec.massMonoisotopic) mp
+    LEFT JOIN
+    (SELECT *, 'Excluded' AS markedAs FROM targetedms.ExcludedPrecursors) exclPrec
```
```suggestion
    FROM targetedms.ExcludedPrecursors exclPrec
```

labkey-bpatel replied:
done!
good suggestion, thanks!

labkey-jeckels commented on resources/queries/targetedms/QCGroupingPrecursorMolecules.sql:
```diff
@@ -0,0 +1,42 @@
+    mp.mz,
+    mp.massMonoisotopic,
+    mp.massAverage,
+    (CASE WHEN exclPrec.markedAs = 'Excluded' THEN 'Excluded' ELSE 'Included' END) AS markedAs
```
```suggestion
    (CASE WHEN exclPrec.RowId IS NULL THEN 'Included' ELSE 'Excluded' END) AS markedAs
```

labkey-jeckels commented on resources/queries/targetedms/QCGroupingPrecursorPeptides.sql:
```diff
@@ -0,0 +1,37 @@
+precPep.mz,
+precPep.neutralMass,
+precPep.name,
+(CASE WHEN exclPrec.markedAs = 'Excluded' THEN 'Excluded' ELSE 'Included' END) AS markedAs
```
Same suggestion as the small molecule variant

labkey-jeckels commented on src/org/labkey/targetedms/outliers/OutlierGenerator.java:
```diff
@@ -344,6 +374,24 @@ private void retainIfPreferred(Map<String, QCMetricConfiguration> preferredConfi
+                                                        Objects.equals(ep.getMassMonoisotopic(), massMonoisotopic) &&
+                                                        Objects.equals(ep.getMassAverage(), massAverage) &&
+                                                        Objects.equals(ep.getCharge(), charge) &&
+                                                        Objects.equals(ep.getMz(), mz)).count() == 1;
```
Here you're looking for exact matches in the numeric values but in the SQL you're rounding to 4 decimal places. Should we be consistent? Can we always use the exact match?

labkey-bpatel replied:
Now using exact match, removed rounding to 4 decimal places. I was trying something out since i wasnt getting expected results, then forgot to remove the rounding.

labkey-jeckels commented on src/org/labkey/targetedms/outliers/OutlierGenerator.java:
```diff
@@ -344,6 +374,24 @@ private void retainIfPreferred(Map<String, QCMetricConfiguration> preferredConfi
 
+    private boolean isExcludedPrecursorPeptide(Collection<ExcludedPrecursor> excludedPrecursors, String modifiedSeq, int charge, double mz)
+    {
+        return excludedPrecursors.stream().filter(ep -> Objects.equals(ep.getModifiedSequence(), modifiedSeq) &&
```
You should be able to use anyMatch() here instead of filter() which would be a tiny bit more efficient. Since the DB doesn't have a constraint to prevent duplicates anymore, this would also handle the case where someone managed to double-insert an exclusion

labkey-bpatel replied:
Good point! Now done

labkey-jeckels reviewed (approved):
A couple of small followup suggestions but no need for additional review unless you want it.

labkey-bpatel committed:
Code review change:
- update custom queries to check for null rowId and remove rounding 4 decimal places
- use anyMatch() instead of filter()

labkey-bpatel committed:
Clean up Excluded Precusors upon folder delete.

labkey-sweta committed:
Automation test : Add Ability to "hide" certain peptides/small molecules (AutoQC)

labkey-bpatel committed:
Fix sqlserver upgrade script file name.
